### PR TITLE
Fix GPU selection issue in the Torch Interface

### LINKED ISSUE #7955: [BUG] pennylane not using correct GPU when using torch
### Expected behavior

The data returned from pennylane's circuit execution should be on the same device as the passed-in parameters. E.g. I prepare a torch array on my cuda:1 GPU and expect the resulting data to also be on cuda:1.

### Actual behavior

The output data is being sent to my default GPU (cuda:0) rather than the expected one (cuda:1).

### Additional information

I am not sure if how I am setting up my QNode is the correct way to do this. I appreciate any advice on changing my code to match the correct intended use.

I am including a simple example, this is not the actual circuit I am interested in but gets the point across.

I am omitting some slight details in machine info from my `qml.about()` call because this is occurring on my work computer but hopefully they are not relevant. Can provide more info if necessary.

### Source code

```shell
import pennylane as qml
import torch

def circuit_builder(n):
    dev = qml.device('default.qubit', wires=n)

    def circuit(params):
        for i in range(n//2):
            qml.CRX(params[i], wires = [2*i, 2*i+1])
        return qml.state()
    
    return qml.QNode(circuit, dev, interface='torch')

print(f'Available GPUs: {torch.cuda.device_count()}')

ckt = circuit_builder(n = 2)

params0 = torch.randn(2, device='cuda:0')
print(f'params0.device: {params0.device}')
state0 = ckt(params0)
print(f'state0.device: {state0.device}')

params1 = torch.randn(2, device='cuda:1')
print(f'params1.device: {params1.device}')
state1 = ckt(params1)
print(f'state1.device: {state1.device}')
```

### Tracebacks

```shell
Output:

Available GPUs: 6
params0.device: cuda:0
state0.device: cuda:0
params1.device: cuda:1
state1.device: cuda:0
```

### System information

```shell
Name: pennylane
Version: 0.41.1
Summary: PennyLane is a cross-platform Python library for quantum computing, quantum machine learning, and quantum chemistry. Train a quantum computer the same way as a neural network.
Home-page:
Author:
Author-email:
License-Expression: Apache-2.0
Location: OMITTED
Requires: appdirs, autograd, autoray, cachetools, diastatic-malt, networkx, numpy, packaging, pennylane-lightning, requests, rustworkx, scipy, tomlkit, typing_extensions
Required-by: PennyLane_Lightning

Platform info:           Linux-5.10....OMITTED
Python version:          3.11.12
Numpy version:           2.3.1
Scipy version:           1.16.0
Installed devices:
- default.clifford (pennylane-0.41.1)
- default.gaussian (pennylane-0.41.1)
- default.mixed (pennylane-0.41.1)
- default.qubit (pennylane-0.41.1)
- default.qutrit (pennylane-0.41.1)
- default.qutrit.mixed (pennylane-0.41.1)
- default.tensor (pennylane-0.41.1)
- null.qubit (pennylane-0.41.1)
- reference.qubit (pennylane-0.41.1)
- lightning.qubit (PennyLane_Lightning-0.41.1)
```

### Existing GitHub issues

- [x] I have searched existing GitHub issues to make sure the issue does not already exist.

### PR DESCRIPTION
**Pull request by maliasadi**
**Context:**
The issue isn't related to the interface impl; limiting the fix to the `qml.math` single and multi dispatchers.

**Benefits:**

1. Fix the GPU selection issue in the `qml.math` wrapper function to coerce a list of (torch) tensors to all have the same dtype in `math/single_dispatch.py`

``` python 
import pennylane as qml
import torch

def circuit_builder(n):
    dev = qml.device('default.qubit', wires=n)

    def circuit(params):
        for i in range(n//2):
            qml.CRX(params[i], wires = [2*i, 2*i+1])
        return qml.state()
    
    return qml.QNode(circuit, dev, interface='torch')

print(f'Available GPUs: {torch.cuda.device_count()}')

ckt = circuit_builder(n = 2)

params0 = torch.randn(2, device='cuda:0')
print(f'params0.device: {params0.device}')
state0 = ckt(params0)
print(f'state0.device: {state0.device}')

params1 = torch.randn(2, device='cuda:1')
print(f'params1.device: {params1.device}')
state1 = ckt(params1)
print(f'state1.device: {state1.device}')
```

Installing PL from `master`: 
```
Available GPUs: 5
params0.device: cuda:0
state0.device: cuda:0
params1.device: cuda:1
state1.device: cuda:0
```

Installing PL with the patch:
```
Available GPUs: 5
params0.device: cuda:0
state0.device: cuda:0
params1.device: cuda:1
state1.device: cuda:1
```

2. Also fix the issue in `qml.math.concatenate`

``` python
import pennylane as qml
import numpy as np

import torch

print(f'Available GPUs: {torch.cuda.device_count()}')

x = torch.tensor([1, 2, 3], device='cuda:2')
y = torch.tensor([4, 5], device='cuda:2')
z = np.array([6, 7, 8])

print(qml.math.concatenate([x, y, z], axis=0))
```

Installing PL from `master`: 
```
Available GPUs: 5
tensor([1, 2, 3, 4, 5, 6, 7, 8], device='cuda:0')
```

Installing PL with the patch:
```
Available GPUs: 5
tensor([1, 2, 3, 4, 5, 6, 7, 8], device='cuda:2')
```

**Possible Drawbacks:**

**Related GitHub Issues:**
Fix #7955
[sc-96197]

### CODE REVIEW

File: pennylane/math/multi_dispatch.py [hunk 1]
```diff
@@ -241,24 +241,40 @@ def concatenate(values, axis=0, like=None):
     if like == "torch":
         import torch
 
-        device = (
-            "cuda"
-            if any(t.device.type == "cuda" for t in values if isinstance(t, torch.Tensor))
-            else "cpu"
-        )
+        device_set = set()
```
Comment (maliasadi) [inline pennylane/math/multi_dispatch.py:244]: `codecov` is failing! Is there any way to check multi-GPU on PL CI, or should we just add `pragma: no cover`?

Comment (JerryChen97) [reply pennylane/math/multi_dispatch.py:244]: have you tried the `gpu` label? Is it helping any way?

Comment (andrijapau) [reply pennylane/math/multi_dispatch.py:244]: Trying now 😄

File: pennylane/math/multi_dispatch.py [hunk 2]
```diff
@@ -241,24 +241,40 @@ def concatenate(values, axis=0, like=None):
...
+        dev_indices = set()
+        torch_device = None
+        for t in values:
+            if isinstance(t, torch.Tensor):
+                device_set.add(t.device.type)
+                dev_indices.add(t.device.index)
+
+        if device_set:  # pragma: no cover
```
Comment (JerryChen97) [inline pennylane/math/multi_dispatch.py:253]:
```suggestion
        if device_set:
```

File: pennylane/math/multi_dispatch.py [hunk 3]
```diff
@@ -241,24 +241,40 @@ def concatenate(values, axis=0, like=None):
...
+
+                raise RuntimeError(
+                    f"Expected all tensors to be on the same device, but found at least two devices, {device_names}!"
+                )
+
+            device = device_set.pop()
+            dev_id = dev_indices.pop() if dev_indices else None
+            torch_device = torch.device(f"{device}:{dev_id}" if dev_id is not None else device)
+
+        else:  # pragma: no cover
```
Comment (JerryChen97) [inline pennylane/math/multi_dispatch.py:266]:
```suggestion
        else:
```

File: pennylane/math/multi_dispatch.py [hunk 4]
```diff
@@ -241,24 +241,40 @@ def concatenate(values, axis=0, like=None):
...
+            torch_device = torch.device("cpu")
 
         if axis is None:
             # flatten and then concatenate zero'th dimension
             # to reproduce numpy's behaviour
             values = [
-                np.flatten(torch.as_tensor(t, device=torch.device(device)))  # pragma: no cover
+                np.flatten(torch.as_tensor(t, device=torch_device))  # pragma: no cover
```
Comment (JerryChen97) [inline pennylane/math/multi_dispatch.py:273]:
```suggestion
                np.flatten(torch.as_tensor(t, device=torch_device))
```

File: pennylane/math/multi_dispatch.py [hunk 5]
```diff
@@ -241,24 +241,40 @@ def concatenate(values, axis=0, like=None):
...
                 for t in values
             ]
             axis = 0
         else:
-            values = [
-                torch.as_tensor(t, device=torch.device(device)) for t in values  # pragma: no cover
-            ]
+            values = [torch.as_tensor(t, device=torch_device) for t in values]  # pragma: no cover
```
Comment (JerryChen97) [inline pennylane/math/multi_dispatch.py:278]:
```suggestion
            values = [torch.as_tensor(t, device=torch_device) for t in values]
```

File: pennylane/math/multi_dispatch.py (hunk 1 above)
Comment (maliasadi) [reply pennylane/math/multi_dispatch.py:244]: Added multi-GPU tests [5d7a421](https://github.com/PennyLaneAI/pennylane/pull/8008/commits/5d7a421e2829f110e15119047e34dfd330b62209)

**Review by andrijapau — APPROVED** [review]
Seems to do what it's promising. Looks good to me!

File: pennylane/math/multi_dispatch.py (hunk 2 above)
Comment (JerryChen97) [inline pennylane/math/multi_dispatch.py:253]: Can we add `TODO`'s here for selfnoting to remove no cover after the multi gpu tester is added in the future?

**Review by JerryChen97 — APPROVED** [review]
THX for fixing!

File: pennylane/math/multi_dispatch.py (hunk 2 above)
Comment (maliasadi) [reply pennylane/math/multi_dispatch.py:253]:
```suggestion
        # TODO: Remove the no-cover pragma once we are able to test with multiple GPUs on CI.
        if device_set:  # pragma: no cover
```
